Add dash widget

zachwe opened the pull request:
Basic version of an "add to dash" widget that allows a user to select from existing plaintext "tags" and create new tags.

jbwyme commented:
I'd let the client handle the data transfer and have the widget be strictly UI focused.

jbwyme commented on lib/widgets/tag-selector/index.js:
```diff
@@ -0,0 +1,98 @@
+    }
+  }
+
+  filterOptions(value) {
```
for internal methods, I've been pseudo-privatizing them by prefixing with `_`: `_filterOptions`. That way they are marked as private and therefore not supported as a public API.

jbwyme commented on lib/widgets/tag-selector/index.js:
```diff
@@ -0,0 +1,98 @@
+  }
+
+  addTag(tagName) {
+    if (this.state.currentTags.indexOf(tagName) !== -1) {
```
how about `if(this.state.currentTags.includes(tagName))`

jbwyme commented on lib/widgets/tag-selector/index.styl:
```diff
@@ -0,0 +1,130 @@
+@import '../../stylesheets/defaults'
+
+tag-styles()
```
any reason for this being a mixin? it looks like it's only used once

zachwe replied:
Yeah it doesn't need to be a mixin as-is. I was initially going to break it into a common mixin location since the dashboard app uses the same tag styles

jbwyme replied:
yeah let's get rid of the mixin

jbwyme commented on lib/widgets/tag-selector/index.js:
```diff
@@ -0,0 +1,98 @@
+  filterOptions(value) {
+    return this.state.options.filter((m) => {
+      return value === '' || m.indexOf(value) !== -1;
+    }).filter((m) => this.state.currentTags.indexOf(m) === -1);
```
I think this could be expressed more simply:

`return this.state.options.filter(option => !this.state.currentTags.includes(option) && option.includes(value));` 

Additionally, we probably want to do a case-insensitive search.

jbwyme commented on lib/widgets/tag-selector/index.styl:
```diff
@@ -0,0 +1,130 @@
+          svg, path
+            color mp-red
+            fill mp-red
+      .arrow-return-icon
```
let's add a few new lines between selectors to break up the wall-o-text

jbwyme commented on lib/widgets/tag-selector/index.js:
```diff
@@ -0,0 +1,98 @@
+      this.update({open});
+      if (open) {
+        setTimeout(() => {
+          this.update({loadState: IDLE});
```
what's this about?

zachwe replied:
yeah I'll get this out of here, was here for simulating like what an http call would look like

zachwe replied:
If we had to do one on widget open as in the designs (I don't think we'll have to at least in the use of this widget on the reports page)

jbwyme commented on lib/widgets/tag-selector/index.js:
```diff
@@ -0,0 +1,98 @@
+    return {
+      defaultState: {
+        currentTags: [],
+        options: [],
```
what are your thoughts on renaming these two properties: `selectedTags`, `allTags`? It might make some of the code below read slightly better.

jbwyme commented on lib/widgets/tag-selector/index.js:
```diff
@@ -0,0 +1,98 @@
+        options: [],
+        value: '',
+        open: true,
+        shouldFocus: true,
```
is this used?

jbwyme commented on lib/widgets/tag-selector/index.js:
```diff
@@ -0,0 +1,98 @@
+      defaultState: {
+        currentTags: [],
+        options: [],
+        value: '',
```
`value` is a bit ambiguous, what about "tagText" or something like that?

jbwyme commented on lib/widgets/tag-selector/index.jade:
```diff
@@ -0,0 +1,44 @@
+      .loader
+        mp-loader
+    else
+      content(select=".slot-header")
```
let's just drop the `select` here. I know we talked about having multiple slots but we don't currently have a design that requires them.

jbwyme commented on lib/widgets/tag-selector/index.jade:
```diff
@@ -0,0 +1,44 @@
+.mp-tag-selector-container
+  mp-drop-menu(attributes={open: open})
```
Assuming we get to place where `open` isn't used internally, can we do `{open: $component.isAttributeEnabled('open')}` and drop `open` from state?

jbwyme commented on lib/widgets/tag-selector/index.js:
```diff
@@ -0,0 +1,98 @@
+        handleTagCloseClick: (tagName) => {
+          this.removeTag(tagName);
+        },
+        handleKeyActions: e => {
```
do we intend on handling the autocomplete options like we do in other places (e.g. the bookmarks-widget that's available in the style-guide now)? Up/down keypress to highlight options, highlighting the characters that match during search, etc

zachwe replied:
Yup good catch, I missed some of those aspects in the design

jbwyme commented on lib/widgets/tag-selector/index.js:
```diff
@@ -0,0 +1,98 @@
+    const currentTags = JSON.parse(this.getAttribute('currentTags'));
+    const options = JSON.parse(this.getAttribute('allTags'));
+    const open = this.getAttribute('open');
+    this.update({currentTags, options, open});
```
nit: let's just do these all inline instead of assigning to var first.

jbwyme commented on lib/widgets/tag-selector/index.styl:
```diff
@@ -0,0 +1,130 @@
+.mp-tag-selector-container
+  border-radius 5px
+  box-shadow: 0 10px 15px 0 rgba(0, 0, 0, 0.1), 0 15px 30px 0 rgba(0, 0, 0, 0.2);
+  width 378px
```
let's make this 100% and let the width of the `host` specify the width. You can use `+host-styles` to give it a default.

jbwyme commented on lib/widgets/tag-selector/index.styl:
```diff
@@ -0,0 +1,130 @@
+    height 72px
+    width 378px
+    display flex
+    mp-loader::shadow
```
you said don't review the loader stuff but just in case: 

1. Shadow DOM v1 removes the `::shadow` selector
2. We shouldn't need to reach into these styles anyways

jbwyme commented on lib/widgets/tag-selector/index.styl:
```diff
@@ -0,0 +1,130 @@
+
+.mp-tag-selector-container
+  border-radius 5px
+  box-shadow: 0 10px 15px 0 rgba(0, 0, 0, 0.1), 0 15px 30px 0 rgba(0, 0, 0, 0.2);
```
can you add `box-sizing border-box` to the container rules? I'm going for consistency on all components and border-box is so much better.

jbwyme commented on lib/widgets/tag-selector/index.styl:
```diff
@@ -0,0 +1,130 @@
+      line-height 1.5
+      padding 8px 0px 10px 9px;
+      &:hover
+        background-color #eff3f5
```
we shouldn't be using any hex values here: `grey-100`

jbwyme commented on lib/widgets/tag-selector/index.styl:
```diff
@@ -0,0 +1,130 @@
+    font-size-medium()
+    padding 14px 96px
+    &:hover
+      background-color #eff3f5
```
`grey-100`

zachwe commented on lib/widgets/tag-selector/index.js:
```diff
@@ -0,0 +1,98 @@
+const LOADING_TAG = 'loading_tag';
+const LOADED_TAG = 'loaded_tag';
+const ERROR = 'error';
+const LOAD_STATES = [LOADING_WIDGET, IDLE, LOADING_TAG, LOADED_TAG, ERROR];
```
I'm basing off the designs—loading widget (zpl.io/2b9JLG), idle (zpl.io/ZXzHA5), loading tag i.e. adding a tag to a report (zpl.io/ZLwLwF), tag loaded success (zpl.io/2wTqjb) and tag load error (the last thing here zpl.io/1jA6bE).

some of these could definitely be parsed out into their own variables but I thought a single state variable would be easier to keep track of

jbwyme replied:
yeah i dont know, im still not sold on this yet but i don't have the right solution in mind so we can revisit once we get closer to the end

zachwe commented on lib/widgets/tag-selector/index.jade:
```diff
@@ -0,0 +1,44 @@
+        for option in $helpers.getSearchMatches()
+          .option(onclick=() => $helpers.addTag(option))
+            | #{option}
+      if $helpers.shouldShowFooter()
```
Seems reasonable

zachwe commented:
I fixed/responded to all your comments, made the style guide example more functional, and implemented text search highlighting (I copied in your `highlight` jade mixin from bookmarks)—that's probably something we want to put in a common location now or later?

Can you take another look?

tdumitrescu commented:
Following up on the bookmarks PR, do we want to pull out the trickier match-finding code from IRB for these two widgets?

jbwyme commented on examples/style-guide/index.js:
```diff
@@ -28,9 +28,16 @@ document.registerElement('style-guide', class extends Component {
         },
         savingBookmark: false,
         selectedBookmarkId: null,
+        showTagSelector: false,
```
is this used?

jbwyme commented on examples/style-guide/index.js:
```diff
@@ -91,6 +98,39 @@ document.registerElement('style-guide', class extends Component {
+          if (e.detail.action === 'addTag') {
+            this.update({tagSelectorLoadState: 'loading_tag'});
+            setTimeout(() => {
+              let allTags = this.state.tagSelectorData.allTags;
```
what if we just make `allTags` and `selectedTags` be sets instead of arrays so you don't even have to check?

jbwyme commented on examples/style-guide/index.js:
```diff
@@ -91,6 +98,39 @@ document.registerElement('style-guide', class extends Component {
+            setTimeout(() => {
+              let allTags = this.state.tagSelectorData.allTags;
+              if (!allTags.includes(tagName)) {
+                allTags = allTags.concat([tagName]);
```
if we _were_ to keep this an array, why not use push instead of concat?

jbwyme commented on examples/style-guide/index.js:
```diff
@@ -91,6 +98,39 @@ document.registerElement('style-guide', class extends Component {
+              this.update({
+                tagSelectorLoadState: 'loaded_tag',
+                tagSelectorData: {
+                  selectedTags: this.state.tagSelectorData.selectedTags.filter((t) => t !== tagName),
```
if we make these sets, you can just use delete: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/delete

jbwyme commented on examples/style-guide/index.js:
```diff
@@ -91,6 +98,39 @@ document.registerElement('style-guide', class extends Component {
+                tagSelectorLoadState: 'loaded_tag',
+                tagSelectorData: {
+                  selectedTags: this.state.tagSelectorData.selectedTags.filter((t) => t !== tagName),
+                  allTags: this.state.tagSelectorData.allTags,
```
? what dis fo?

jbwyme commented on lib/stylesheets/mixins/icons.styl:
```diff
@@ -49,3 +51,24 @@ icon-input(icon-color=grey-500, icon-color-focus=grey-700)
     svg-icon-style(icon-color)
+
+// Much of these styles taken from https://github.com/lukehaas/css-loaders (MIT license)
+loader-styles(size=10px, time=1.1s, spinner-color=mp-blue, background-color=white, spinner-width=3px)
```
we don't need this anymore right?

zachwe replied:
yeah it's because this was rebased off the old loader commit, i'll rebase master

jbwyme commented on lib/widgets/tag-selector/index.jade:
```diff
@@ -0,0 +1,54 @@
+mixin highlight(target, search, highlightClass)
```
we should put this mixin in some common file and use it in this and bookmarks-widget, maybe `lib/widgets/jade/mixins.jade` ? Open to debate on the folder structure.

jbwyme commented on lib/widgets/tag-selector/index.jade:
```diff
@@ -0,0 +1,54 @@
+                svg-icon(
+                  attributes={icon: "x"}
+                  class="close-icon"
+                  onclick=() => $helpers.handleTagCloseClick(tag)
```
i'd just call this `removeTag` instead of `handleTagCloseClick`, more descriptive of the action at the call site.

jbwyme commented on lib/widgets/tag-selector/index.jade:
```diff
@@ -0,0 +1,54 @@
+                )
+          input(
+            class="search-input"
+            oninput=$helpers.handleSearchInput
```
`s/handleSearchInput/setInputText/`, similar to the above, if we know the action we're going to take it's best to be descriptive at the call site

jbwyme commented on lib/widgets/tag-selector/index.jade:
```diff
@@ -0,0 +1,54 @@
+            onkeydown=$helpers.handleKeyActions
+            type="text"
+          )
+          .border(class= loadState == 'error' ? 'loaded-error' : '')
```
`.border(class={'loaded-error': loadState === 'error'})`

jbwyme commented on lib/widgets/tag-selector/index.jade:
```diff
@@ -0,0 +1,54 @@
+        for option, i in $helpers.getSearchMatches()
+          .option(
+            onclick=() => $helpers.addTag(option)
+            class={'active': i == 0 && inputText !== ''}
```
`===`

jbwyme commented on lib/widgets/tag-selector/index.js:
```diff
@@ -0,0 +1,91 @@
+  attributeChangedCallback(attr) {
+    super.attributeChangedCallback(...arguments);
+    if (attr === 'open') {
+      this.update();
```
don't need to do this as it already happens for you: https://github.com/mixpanel/panel/blob/master/lib/component.js#L261

jbwyme commented on lib/widgets/tag-selector/index.js:
```diff
@@ -0,0 +1,91 @@
+    super.attachedCallback(...arguments);
+    this.update({
+      selectedTags: JSON.parse(this.getAttribute('selected-tags')),
+      allTags: JSON.parse(this.getAttribute('all-tags')),
```
what if a load-state is set before it's attached? doesn't look like we handle it.

zachwe replied:
added below

jbwyme commented on lib/widgets/tag-selector/index.js:
```diff
@@ -0,0 +1,91 @@
+
+  attributeChangedCallback(attr) {
+    super.attributeChangedCallback(...arguments);
+    if (attr === 'open') {
```
make this a switch?

jbwyme commented on lib/widgets/tag-selector/index.js:
```diff
@@ -0,0 +1,91 @@
+  }
+
+  _filterOptions(inputText) {
+    return this.state.allTags.filter(tag => !this.state.selectedTags.includes(tag) && tag.toLowerCase().includes(inputText.toLowerCase()));
```
multiline

jbwyme commented on lib/widgets/tag-selector/index.styl:
```diff
@@ -0,0 +1,162 @@
+      font-weight weight-medium
+      height 18px
+      line-height 1.5
+      padding 8px 0px 10px 9px;
```
remove `;`

jbwyme commented on lib/widgets/tag-selector/index.js:
```diff
@@ -0,0 +1,91 @@
+        },
+        handleKeyActions: e => {
+          if (e.keyCode === 13) { // enter
+            this._addTag(this.state.inputText);
```
shouldn't this select the highlighted item if there is one? http://cometapp.us/c/eTzfBUuv

jbwyme replied:
also the footer should reset after _addTag is called shouldn't it? http://cometapp.us/c/REcJrxCq

zachwe replied:
Yup, agree on both of these

jbwyme commented on lib/widgets/tag-selector/index.js:
```diff
@@ -0,0 +1,91 @@
+            if (this.state.inputText.length === 0) {
+              this._removeTag(this.state.selectedTags[this.state.selectedTags.length - 1]);
+            }
+          }
```
what about up/down arrow keys for navigating the options?

jbwyme commented on lib/widgets/tag-selector/index.jade:
```diff
@@ -0,0 +1,54 @@
+          if loadState == 'loading_tag'
+            mp-loader
+          else if loadState == 'loaded_tag'
+            svg-icon(attributes={icon: "tag"} class="tag-icon loaded")
```
this behavior seems useless to me since it stays green once you add one tag. It'd make sense if it showed green and then transitioned back to grey.

zachwe replied:
Yeah this is in the designs but idk what the optimal behavior is. Maybe it should become grey again once the user starts typing

jbwyme commented:
@tdumitrescu yes we *do* want to do that. I was mulling it over last night about how to make it generic and at what level of abstraction it would exist. do you have a proposal for how we might implement the behavior generically?

tdumitrescu commented:
I'm not sure of what the requirements are for autocomplete, and whether they're different from other searches. Here's the jsdoc of the matching function from IRB for a single string + single set of search term input, maybe you can judge whether it would be useful:
```js
/**
 * Split a string into match and non-match substrings based on finding
 * all whitespace-separated terms in a given filter string (for UI search
 * bars).
 * @param {string} str - string to search
 * @param {string} filterStr - string with filter/search terms
 * @returns {Array} list of alternating matching and non-matching substrings
 * in order of the original string; even = match, odd = no match
 * @example
 * stringFilterMatches('my example string', 'ex my');
 * // ['my', ' ', 'ex', 'ample string']
 */
```
Basically in a UI, each item in the returned array becomes a `<span>` with alternating match/non-match classes for styling. Here are the test cases:
```js
describe('stringFilterMatches', function() {
  it('matches at the beginning of a string', function() {
    expect(stringFilterMatches('abcdefg', 'abc')).to.eql(['abc', 'defg']);
  });

  it('matches in the middle/end of a string', function() {
    expect(stringFilterMatches('abcdefg', 'def')).to.eql(['', 'abc', 'def', 'g']);
    expect(stringFilterMatches('abcdefg', 'efg')).to.eql(['', 'abcd', 'efg', '']);
  });

  it('returns null for non-matches', function() {
    expect(stringFilterMatches('abcdefg', 'ac')).to.eql(null);
  });

  it('is case-insensitive', function() {
    expect(stringFilterMatches('abcdefg', 'dEf')).to.eql(['', 'abc', 'def', 'g']);
    expect(stringFilterMatches('abCDEfg', 'dEF')).to.eql(['', 'abC', 'DEf', 'g']);
  });

  it('ignores whitespace padding', function() {
    expect(stringFilterMatches('abcdefg', '   def')).to.eql(['', 'abc', 'def', 'g']);
    expect(stringFilterMatches('abcdefg', 'abc   ')).to.eql(['abc', 'defg']);
    expect(stringFilterMatches('abcdefg', '  abc ')).to.eql(['abc', 'defg']);
  });

  it('matches when no filter string is passed', function() {
    expect(stringFilterMatches('abcdefg', '')).to.eql(['', 'abcdefg']);
    expect(stringFilterMatches('abcdefg', null)).to.eql(['', 'abcdefg']);
  });

  it('matches when all space-separated terms match', function() {
    expect(stringFilterMatches('abcdefg', 'abc efg')).to.eql(['abc', 'd', 'efg', '']);
    expect(stringFilterMatches('abcdefg', 'abc   efg')).to.eql(['abc', 'd', 'efg', '']);
    expect(stringFilterMatches('abcdefg', '  abc   efg   ')).to.eql(['abc', 'd', 'efg', '']);
  });

  it('matches out of order', function() {
    expect(stringFilterMatches('abcdefg', 'efg abc')).to.eql(['abc', 'd', 'efg', '']);
  });

  it('merges contiguous matches', function() {
    expect(stringFilterMatches('abcdefg', 'efg bcd')).to.eql(['', 'a', 'bcdefg', '']);
  });

  it('does not match when one or more space-separated terms do not match', function() {
    expect(stringFilterMatches('abcdefg', 'abc xxx')).to.eql(null);
  });
});
```

jbwyme commented:
haha nice that's basically what we did with this mixin: https://github.com/mixpanel/mixpanel-common/pull/94/commits/7e2ed7fdbaf38a1d92b5b9c4891a5a8a315c8d83#diff-3f260d6e12ca0d660863f2567684da01R1

tdumitrescu commented:
Yup, we got a jade mixin for it in irb too.

jbwyme commented:
well start contributing your shit to mixpanel-common homie.

On Wed, Dec 14, 2016 at 11:21 PM, tdumitrescu <notifications@github.com>
wrote:

> Yup, we got a jade mixin for it in irb too.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/mixpanel/mixpanel-common/pull/94#issuecomment-267241709>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAh_UqHOmDzfjWeYOKf2RGtLFfdSZcaOks5rIM5xgaJpZM4LIXUE>
> .
>

tdumitrescu commented:
So far it hasn't been needed elsewhere. I offered when I saw you doing filtering in the BMW!

jbwyme commented on lib/widgets/tag-selector/index.js:
```diff
@@ -0,0 +1,91 @@
+  }
+
+  _addTag(tagName) {
+    if (this.state.selectedTags.includes(tagName)) {
```
let's just make this one block
```
if (!this.state.selectedTags.incudes(tagName)) {
  this.setAttribute......
}
```

jbwyme commented on lib/widgets/tag-selector/index.styl:
```diff
@@ -0,0 +1,155 @@
+
++host-styles(mp-drop-menu)
+  display block
+  width 378px
```
how does this width and the one above relate to each other?

jbwyme commented on lib/widgets/tag-selector/index.js:
```diff
@@ -37,7 +37,7 @@ registerMPElement('mp-tag-selector', class extends Component {
         },
         handleKeyActions: e => {
-          if (e.keyCode === 13) { // enter
+          if (e.key === 'Enter') {
```
why'd you switch to this?

zachwe replied:
It looks like `keyCode` is deprecated (https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode), and it wasn't being set in tests even if passed into the EventInit object

jbwyme replied:
support for it may be deprecated but it's not dropped and `key` is not fully supported by a lot of browsers yet (and not at all in Safari): https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key#Browser_compatibility

jbwyme commented on test/widgets/tag-selector.js:
```diff
@@ -0,0 +1,51 @@
+const selectedTags = ['tag foo'];
+
+describe('Test tag-selector widget', () => {
+  beforeEach(function(done) {
```
no need for `done` if there's no async code in all of these functions

jbwyme commented on test/widgets/tag-selector.js:
```diff
@@ -0,0 +1,51 @@
+
+  it('fires change event to add tags', function(done) {
+    this.widget.addEventListener('change', e => {
+      console.log(Object.keys(e));
```
remove logs

jbwyme commented on lib/widgets/tag-selector/index.jade:
```diff
@@ -0,0 +1,54 @@
+    span(class={[highlightClass]: token.toLowerCase() === search.toLowerCase()})= token
+
+.mp-tag-selector-container
+  mp-drop-menu(attributes={open: $component.isAttributeEnabled('open')})
```
how much work would it be to get this out of here like we talked about last week? I'd prefer not to merge this with the drop-menu if we can avoid it.

zachwe replied:
removed in https://github.com/mixpanel/mixpanel-common/pull/94/commits/a7f02fe7a1dfff619c8661ccc57b9fb08fc63944

jbwyme commented on test/widgets/tag-selector.js:
```diff
@@ -0,0 +1,45 @@
+  });
+
+  it('fires change event to add tags', function() {
+    this.widget.addEventListener('change', e => {
```
this method *is* async so it needs the `done` behavior

jbwyme commented on test/widgets/tag-selector.js:
```diff
@@ -0,0 +1,45 @@
+  });
+
+  it('fires change event to remove tags', function() {
+    this.widget.addEventListener('change', e => {
```
same with this one

zachwe commented:
doh, fixed and rebased + updated a bunch of stuff to pass new linter rules
